<?xml version="1.0"?>

<robot name="diff_drive">
    <gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
            <robotNamespace>/diff_drive</robotNamespace>
        </plugin>
    </gazebo>
    <link name="chassis">
        <visual>
        <origin xyz="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.6 0.4 0.2"/>
            </geometry>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0"/>
            <geometry>
                <box size="0.6 0.4 0.2"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="40"/>
            <inertia
                ixx="0.33" ixy="0.0" ixz="0.0"
                iyy="0.33" iyz="0.0"
                izz="0.33"/>
        </inertial>
    </link>

    <link name="left_wheel">
        <visual>
            <origin xyz="0 0 0" rpy="1.57 0 3.14"/>
            <geometry>
                <cylinder radius="0.2" length="0.1"/>
            </geometry>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="1.57 0 3.14"/>
            <geometry>
                <cylinder radius="0.2" length="0.1"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="10"/>
            <inertia ixx="0.2" ixy="0" ixz="0" iyy="0.2" iyz="0" izz="0.2"/>
        </inertial>
    </link>
    <joint name="chassis_to_left_wheel" type="continuous">
        <origin xyz="-0.1 0.25 0"/>
        <axis xyz="0 1.0 0.0"/>
        <parent link="chassis"/>
        <child link="left_wheel"/>
    </joint>

    <transmission name="left_wheel_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="chassis_to_left_wheel">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </joint>
        <actuator name="left_wheel_motor">
            <mechanicalReduction>1</mechanicalReduction>
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </actuator>
    </transmission>


    <link name="right_wheel">
        <visual>
            <origin xyz="0 0 0" rpy="1.57 0 0"/>
            <geometry>
                <cylinder radius="0.2" length="0.1"/>
            </geometry>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="1.57 0 0"/>
            <geometry>
                <cylinder radius="0.2" length="0.1"/>
            </geometry>
        </collision>
        <inertial>
            <mass value="10"/>
            <inertia ixx="0.2" ixy="0" ixz="0" iyy="0.2" iyz="0" izz="0.2"/>
        </inertial>
    </link>
    <joint name="chassis_to_right_wheel" type="continuous">
        <origin xyz="-0.1 -0.25 0"/>
        <axis xyz="0.0 1.0 0.0"/>
        <parent link="chassis"/>
        <child link="right_wheel"/>
    </joint>

    <transmission name="right_wheel_trans">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="chassis_to_right_wheel">
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </joint>
        <actuator name="right_wheel_motor">
            <mechanicalReduction>1</mechanicalReduction>
            <hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
        </actuator>
    </transmission>

    <link name="caster_wheel">
        <inertial>
            <origin xyz="0.0 0.0 0.0"/>
            <mass value="25.0"/>
            <inertia ixx="1" ixy="0" ixz="0" iyy="1" iyz="0.0" izz="1.0"/>
        </inertial>
        <visual>
            <origin xyz="0.0 0.0 0.0"/>
            <geometry>
                <sphere radius="0.05"/>
            </geometry>
        </visual>
        <collision>
            <origin xyz="0.0 0.0 0.0"/>
            <geometry>
                <sphere radius="0.05"/>
            </geometry>
        </collision>
    </link>
    <joint name="chassis_to_caster_wheel" type="continuous">
        <parent link="chassis"/>
        <child link="caster_wheel"/>
        <origin xyz="0.2 0.0 -0.15"/>
        <axis xyz="0.0 1.0 0.0"/>
    </joint>



</robot>